src/wp-includes/option.php

```diff
@@ -805,7 +805,13 @@ function update_option( $option, $value, $autoload = null ) {
 	 *
 	 * See https://core.trac.wordpress.org/ticket/38903 and https://core.trac.wordpress.org/ticket/22192.
 	 */
-	if ( $raw_old_value !== $default_value && _is_equal_database_value( $raw_old_value, $value ) ) {
+	if (
+		$value === $raw_old_value ||
```

Comment:
I am not sure why this is needed, isnt this handled by `_is_equal_database_value`?

Reply:
We got one BC issue [raise here](https://core.trac.wordpress.org/ticket/22192#comment:52) and that we addressed in [changeset 56717](https://core.trac.wordpress.org/changeset/56717). Hope that commit message answers your queries.

Reply:
Okay, what about 

```
_is_equal_database_value( $raw_old_value, $value ) || $raw_old_value !== $default_value && 
```

Is that the same?

https://github.com/WordPress/wordpress-develop/blob/8bbeee43b647d6c8fc54c67cb0616343ac7b03fb/src/wp-includes/option.php#L2934-L2936

Reply:
The bail cases are:
- The old and new values are strictly equal.
- The old and new values are loosely equal AND the old value is not the default value. - Note that this case has two conditions.

Reply:
Maybe we should rename `_is_equal_database_value` to say loosely equal, to make this easier to read then.

Reply:
I am not convinced that this additional clause here is needed either.

I have read through https://github.com/WordPress/wordpress-develop/pull/5272#issuecomment-1740347838, and I acknowledge that technically storing the option if the given value is the same as the default is a backward compatibility break. However, we have also previously clarified that _not_ allowing that leads to another problem because the `get_option()` result of this scenario would in fact be different, so it _should_ set the value in the database:
* For a non-existent option, `get_option()` returns `false`.
* For an option stored with `update_option( 'foo', false )`, `get_option()` returns `''`.

See for example @joemcgill's comment in https://github.com/WordPress/wordpress-develop/pull/5139#discussion_r1323340693: The above example leads to a different return type value, so I think it should result in an update.

This is a crucial differentiation. Unless you have encountered this BC break in real-world examples, I personally think this BC break is an edge case that is okay to accept, for two reasons:
* I cannot imagine that someone calls `update_option( 'foo', false )` and explicitly wants that value to not be stored in the DB.
* Regarding the return value, the majority of plugins don't care about the return value of `update_option()` and assume it works as expected.

Reply:
@felixarntz Since this ticket is about reducing queries by comparing values, I don't think creating a new option in the database when `old value === new value === default value` is within that scope, and such a change would likely need its own commit for better tracking of history.

This strict check prevents a an unnecessarily and unplanned regression/change to `update_option()`'s behaviour.

Reply:
@costdev I _may_ agree with that.

Main point though is that we should decouple the production code change here from the tests added, since by far most tests here were just for improved coverage of the original change, so the fix here, whether it's committed or not, should then also go in on its own IMO.